### ISSUE TITLE
False negative in RequireThis check

### ISSUE THREAD

**Issue by Vladlis** [checkstyle/checkstyle#2240]
Input:
```java
class Test {
    private int foo;

    void bar() {
        foo++;
        int foo;
    }       
}
```
Config:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <module name="TreeWalker">
         <module name="RequireThis"/>
    </module>
</module>
```
Result:
```bash
vlad@vlad-N61PC-M2S ~/cli $ java -jar checkstyle-6.10.1-all.jar -c my_check.xml Test.java
Starting audit...
Audit done.
```
The check should require ```this``` for ```foo```.
The problem is in check's superclass: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/AbstractDeclarationCollector.java#L193 - it searches by name without checking the line of variable declaration.

**Comment by Vladlis** [comment]
Another false-negative:
Input
```java
public class Test {
	void bar() {
		int foo;
		foo();             // <-- violation expected
	}
	
	void foo() {
	}
}
```
Config:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <module name="TreeWalker">
         <module name="RequireThis"/>
    </module>
</module>
```
Result:
```
vlad@vlad-N61PC-M2S ~/cli $ java -jar checkstyle-6.12.1-all.jar -c my_check.xml Test.java
Starting audit...
Audit done.
```
